java-checks/src/main/java/org/sonar/java/checks/helpers/HardcodedStringExpressionChecker.java

```diff
@@ -0,0 +1,212 @@
+      case METHOD_INVOCATION:
+        MethodInvocationTree methodInvocationTree = (MethodInvocationTree) arg;
+        return isDerivedFromPlainText(methodInvocationTree, secondaryLocations, visited);
+      case CONDITIONAL_EXPRESSION: 
+        ConditionalExpressionTree conditionalTree = (ConditionalExpressionTree) arg;
+        return isDerivedFromPlainText(conditionalTree, secondaryLocations, visited);
+      case MEMBER_SELECT: 
+        MemberSelectExpressionTree memberSelect = (MemberSelectExpressionTree) arg;
+        return isDerivedFromPlainText(memberSelect.identifier(), secondaryLocations, visited);
+      case STRING_LITERAL:
+        return !LiteralUtils.isEmptyString(arg);
+      case TYPE_CAST:
+        TypeCastTree typeCast = (TypeCastTree) arg;
+        return isExpressionDerivedFromPlainText(typeCast.expression(), secondaryLocations, visited);
+      case BOOLEAN_LITERAL:
+      case CHAR_LITERAL:
+      case DOUBLE_LITERAL:
+      case FLOAT_LITERAL:
+      case INT_LITERAL:
+      case LONG_LITERAL:
+        return true;
+      default:
+        if (arg instanceof BinaryExpressionTree) {
+          BinaryExpressionTree binaryExpression = (BinaryExpressionTree) arg;
+          return isDerivedFromPlainText(binaryExpression, secondaryLocations, visited);
+        }
+        return false;
+    }
+  }
+
+  private static boolean isDerivedFromPlainText(BinaryExpressionTree binaryExpression, List<JavaFileScannerContext.Location> secondaryLocations,
+    Set<Symbol> visited) {
+    return isExpressionDerivedFromPlainText(binaryExpression.rightOperand(), secondaryLocations, visited) &&
+      isExpressionDerivedFromPlainText(binaryExpression.leftOperand(), secondaryLocations, visited);
+  }
+
+  private static boolean isDerivedFromPlainText(IdentifierTree identifier, List<JavaFileScannerContext.Location> secondaryLocations,
+    Set<Symbol> visited) {
+    Symbol symbol = identifier.symbol();
+    boolean firstVisit = visited.add(symbol);
+    if (!firstVisit || !symbol.isVariableSymbol() || JUtils.isParameter(symbol) || isNonFinalField(symbol)) {
+      return false;
+    }
+    VariableTree variable = (VariableTree) symbol.declaration();
+    if (variable == null) {
+      return JUtils.constantValue((Symbol.VariableSymbol) symbol).isPresent();
+    }
+
+    List<ExpressionTree> assignments = getIdentifierAssignments(identifier);
+
+    List<JavaFileScannerContext.Location> tempSecondaryLocations = new ArrayList<>();
+    boolean identifierIsDerivedFromPlainText = !assignments.isEmpty() &&
+      assignments.stream()
+        .allMatch(expression -> isExpressionDerivedFromPlainText(expression, tempSecondaryLocations, visited));
+
+    if (identifierIsDerivedFromPlainText) {
+      if(variable.initializer() == null) {
+        secondaryLocations.add(new JavaFileScannerContext.Location(SECONDARY_LOCATION_ISSUE_MESSAGE, variable));        
+      }else {
```

Comment:
looks like styling issue
```suggestion
      if (variable.initializer() == null) {
        secondaryLocations.add(new JavaFileScannerContext.Location(SECONDARY_LOCATION_ISSUE_MESSAGE, variable));        
      } else {
```